Suppress highlighting inside strings for modes which have them

wi11dey opened the pull request:
Thanks for writing this.

This PR makes it so strings like `"asdf ;;; jkl"` are not highlighted--only comments are. Previously, text inside strings in programming modes that matched `outline-minor-faces-regexp` were highlighted to the end of the line if your regexp was not anchored to beginning-of-line. For example, my regexp is `";;;+[^#].*$"` so indented comments can also be headings, but it results in:

![previous](https://user-images.githubusercontent.com/52264243/77851782-76483e80-71ca-11ea-83c8-62d312446f81.png)

This made for some confusing syntax highlighting, and is a problem with the default regexp for multiline strings.

Now, everything that used to be highlighted still is, except for matches inside strings.

The way it works is that if `font-lock-keywords-only` is `nil` (true for modes that have strings and comments), a function is used as the matcher which checks if the matched text has already been fontified as a string, and returns `nil` if so. If keyword-only matching is on (true for text modes, etc.) then the regexp is used directly and there is no performance hit.

I've also changed `outline-minor-faces--font-lock-keywords` to a static variable rather than a function, since according to the [manual](https://www.gnu.org/software/emacs/manual/html_node/elisp/Search_002dbased-Fontification.html#Search_002dbased-Fontification), Font Lock keywords with `eval` are evaluated only _once_ in a buffer so the effect is the same as running a function to get the matcher while being simpler.

tarsius commented:
Ha! I have been working on exactly the same code this afternoon...

tarsius commented:
> I've also changed outline-minor-faces--font-lock-keywords to a static variable

That should be done in a separate commit.

tarsius commented on outline-minor-faces.el:
```diff
@@ -105,7 +105,11 @@ be used directly because it is only supposed to match the
 
 (defun outline-minor-faces--syntactic-matcher (regexp)
-  "Returns a matcher that matches REGEXP only if it is not found inside a string. Returns REGEXP directly for modes where `font-lock-keywords-only' is non-nil because Font Lock does not mark strings and comments for those modes, and the matcher will not know what is/is not a string. "
+  "Returns a matcher that matches REGEXP only if it is not found
```
The first sentence must completely fit on the first line. Maybe "Return a matcher that matches REGEXP only outside of strings." Generally if such a first (aka summary) sentence is not precise enough, then that information should be repeated and augmented in the text that follows, even if that leads to some duplicated information.

And there's no need to do this in a separate commit, just squash into the appropriate existing commit. Rewriting "history" (aka it's not actually history yet) is perfectly okay on feature branches, even when they are public.

wi11dey committed:
Suppress highlighting inside strings for modes which have them

wi11dey committed:
Switch function+backquote to unchanging variable

According to the manual, Font Lock keywords with `eval` are evaluated
only once in a buffer so the effect is the same as running a function
to get the matcher while being simpler.

wi11dey commented:
Thanks for the quick review! I've re-done the commits starting from master so that there's one for new functionality and one for function->var. Docstring should also be up to convention now

tarsius commented:
I've fixed how strings and comments are detected. Your implementation did not work for doc-strings at least.

tarsius commented:
Thanks!
